nixos/modules/virtualisation/ec2-metadata-fetcher.nix

```diff
@@ -1,23 +1,37 @@
-{ targetRoot, wgetExtraOptions }:
+# Note: This fetcher is used by both ./amazon-image.nix and
+# ./openstack-config.nix, so we need to be compatible with both.
+#
+# More info about OpenStack's IMDS, including version compatibility:
+# https://docs.openstack.org/nova/latest/user/metadata.html#metadata-ec2-format
+{ targetRoot, wgetExtraOptions, useImdsToken ? false }:
+let
+  # Snippet to fetch the IMDS token, used by IMDSv2:
+  # https://aws.amazon.com/blogs/security/defense-in-depth-open-firewalls-reverse-proxies-ssrf-vulnerabilities-ec2-instance-metadata-service/
+  # We construct the PUT request by hand, because wget is busybox wget and can't.
+  wgetCmdWithImdsToken = ''
+    TOKEN=$(echo -ne 'PUT /latest/api/token HTTP/1.1\r\nHost: 169.254.169.254\r\nX-aws-ec2-metadata-token-ttl-seconds: 60\r\n\r\n' | nc 169.254.169.254 80 | tail -n 1)
+    function wgetCmd() {
+      wget ${wgetExtraOptions} --header "X-aws-ec2-metadata-token: $TOKEN" "$@";
+    }
+  '';
+  wgetCmdWithoutImdsToken = ''
+    function wgetCmd() {
+      wget ${wgetExtraOptions} "$@";
+    }
+  '';
+in
 ''
+  imds=http://169.254.169.254/2009-04-04
   metaDir=${targetRoot}etc/ec2-metadata
   mkdir -m 0755 -p "$metaDir"
 
   echo "getting EC2 instance metadata..."
 
-  if ! [ -e "$metaDir/ami-manifest-path" ]; then
-    wget ${wgetExtraOptions} -O "$metaDir/ami-manifest-path" http://169.254.169.254/1.0/meta-data/ami-manifest-path
-  fi
+  ${if useImdsToken then wgetCmdWithImdsToken else wgetCmdWithoutImdsToken}
 
-  if ! [ -e "$metaDir/user-data" ]; then
-    wget ${wgetExtraOptions} -O "$metaDir/user-data" http://169.254.169.254/1.0/user-data && chmod 600 "$metaDir/user-data"
-  fi
-
-  if ! [ -e "$metaDir/hostname" ]; then
-    wget ${wgetExtraOptions} -O "$metaDir/hostname" http://169.254.169.254/1.0/meta-data/hostname
-  fi
-
-  if ! [ -e "$metaDir/public-keys-0-openssh-key" ]; then
-    wget ${wgetExtraOptions} -O "$metaDir/public-keys-0-openssh-key" http://169.254.169.254/1.0/meta-data/public-keys/0/openssh-key
-  fi
```

Comment:
Should I be doing a `rm -f $metaDir/*` here? Otherwise, files that might exist in certain conditions only (e.g. `/etc/ec2-metadata/public-ipv4`) might persist when they should be removed (if the public ipv4 goes away for some reason).